File: ionic/mrgenie/www/todoApp/js/services.js
angular.module('todoApp.services', [])
    .factory('Todo', ['$http', 'PARSE_CREDENTIALS', function ($http, PARSE_CREDENTIALS) {
        return {
            getAll: function () {
                return $http.get('https://api.parse.com/1/classes/Todo', {
                    headers: {
                        'X-Parse-Application-Id': PARSE_CREDENTIALS.APP_ID,
                        'X-Parse-REST-API-Key': PARSE_CREDENTIALS.REST_API_KEY,
                    }
                });
            },
            get: function (id) {
                return $http.get('https://api.parse.com/1/classes/Todo/' + id, {
                    headers: {
                        'X-Parse-Application-Id': PARSE_CREDENTIALS.APP_ID,
                        'X-Parse-REST-API-Key': PARSE_CREDENTIALS.REST_API_KEY,
                    }
                });
            },
            create: function (data) {
                return $http.post('https://api.parse.com/1/classes/Todo', data, {
                    headers: {
                        'X-Parse-Application-Id': PARSE_CREDENTIALS.APP_ID,
                        'X-Parse-REST-API-Key': PARSE_CREDENTIALS.REST_API_KEY,
                        'Content-Type': 'application/json'
                    }
                });
            },
            edit: function (id, data) {
                return $http.put('https://api.parse.com/1/classes/Todo/' + id, data, {
                    headers: {
                        'X-Parse-Application-Id': PARSE_CREDENTIALS.APP_ID,
                        'X-Parse-REST-API-Key': PARSE_CREDENTIALS.REST_API_KEY,
                        'Content-Type': 'application/json'
                    }
                });
            },
            delete: function (id) {
                return $http.delete('https://api.parse.com/1/classes/Todo/' + id, {
                    headers: {
                        'X-Parse-Application-Id': PARSE_CREDENTIALS.APP_ID,
                        'X-Parse-REST-API-Key': PARSE_CREDENTIALS.REST_API_KEY,
                        'Content-Type': 'application/json'
                    }
                });
            }
        }
    }])
    .value('PARSE_CREDENTIALS', {
        APP_ID: 'tdhNUXe5ytcL28D4e0hv5782fFw0N6iFkIS7Tiyn',
        REST_API_KEY: 'aoi46j5L12B2o1pNmGzceCSTj3cyzgGb5nK8uPzA'
    })
;

File: python/mrgenie/mrgenie/parsecom/services.py
from datetime import datetime
from datetime import timedelta
import json

import urllib3
from mrgenie.services import Service
import settings


def urlencode(params):
    return '&'.join(['{}={}'.format(k, json.dumps(v).replace(' ', ''))
                     for k, v in params.items()])


def get(path, params=None):
    if not params:
        params = {}

    params_urlencoded = urlencode(params)

    if settings.PROXY:
        http = urllib3.ProxyManager(settings.PROXY)
    else:
        http = urllib3.PoolManager()

    resp = http.request(
        'GET',
        'https://api.parse.com/{}?{}'.format(path, params_urlencoded),
        headers={
            "X-Parse-Application-Id": settings.PARSE_APP_ID,
            "X-Parse-REST-API-Key": settings.PARSE_REST_KEY
        }
    )
    return resp.data.decode()


def post_or_put(path, params=None, method='POST'):
    if not params:
        params = {}

    if settings.PROXY:
        http = urllib3.ProxyManager(settings.PROXY)
    else:
        http = urllib3.PoolManager()

    resp = http.request(
        method,
        'https://api.parse.com/' + path,
        headers={
            "X-Parse-Application-Id": settings.PARSE_APP_ID,
            "X-Parse-REST-API-Key": settings.PARSE_REST_KEY
        },
        body=json.dumps(params)
    )
    return resp.data.decode()


def post(path, params=None):
    return post_or_put(path, params)


def put(path, params=None):
    return post_or_put(path, params, method='PUT')


def delete(path):
    if settings.PROXY:
        http = urllib3.ProxyManager(settings.PROXY)
    else:
        http = urllib3.PoolManager()

    resp = http.request(
        'DELETE',
        'https://api.parse.com/' + path,
        headers={
            "X-Parse-Application-Id": settings.PARSE_APP_ID,
            "X-Parse-REST-API-Key": settings.PARSE_REST_KEY
        }
    )
    return resp.decode()


def to_parse_date(date):
    parse_date = datetime(date.year, date.month, date.day, date.hour - 2, date.minute)
    return {
        "__type": "Date",
        "iso": parse_date.strftime('%Y-%m-%dT%H:%M:%S.%fZ')
    }


class ParseService(Service):
    def get_rooms(self):
        rooms_json = json.loads(get(path='/1/classes/Room'))
        rooms = rooms_json['results']
        return [(x['objectId'], x['name']) for x in rooms]

    def get_reservations(self, room_id):
        path = '/1/classes/Reservation'
        params = {
            'where': {
                'room': {
                    '__type': 'Pointer',
                    'className': 'Room',
                    'objectId': room_id
                }
            }
        }

        reservations_raw = json.loads(get(path, params))['results']

        reservations = [
            {
                'start_date': to_date(x['start_date']['iso']),
                'end_date': to_date(x['end_date']['iso']),
                'objectId': x['objectId'],
            } for x in reservations_raw]

        return reservations

    def get_current_reservation_id(self, room_id):
        start_date = datetime.now()
        reservations = self.get_reservations(room_id)
        for reservation in reservations:
            if reservation['start_date'] < start_date < reservation['end_date']:
                return reservation['objectId']

    def make_reservation(self, room_id):
        if self.get_current_reservation_id(room_id):
            return False

        start_date = datetime.now()
        end_date = start_date + timedelta(minutes=30)
        params = {
            'start_date': to_parse_date(start_date),
            'end_date': to_parse_date(end_date)
        }

        path = '/1/classes/Reservation'
        resp = post(path=path, params=params)
        if resp:
            reservation_object_id = json.loads(resp)['objectId']
            reservation_path = path + '/' + reservation_object_id
            rel_params = {
                'room': {
                    "__op": "AddRelation",
                    "objects": [
                        {
                            "__type": "Pointer",
                            "className": "Room",
                            "objectId": room_id
                        }
                    ]
                }
            }

            put(path=reservation_path, params=rel_params)

        return resp

    def cancel_reservation(self, room_id):
        reservation_id = self.get_current_reservation_id(room_id)
        if reservation_id:
            delete('/1/classes/Reservation/' + reservation_id)
            return True
        return False


def to_date(strdate):
    parse_date = datetime.strptime(strdate[:19], "%Y-%m-%dT%H:%M:%S")
    date = datetime(parse_date.year, parse_date.month, parse_date.day, parse_date.hour + 2, parse_date.minute)
    return date
